### PR TITLE
DisallowMultipleAttributesPerLineSniff

### PR DESCRIPTION
**Pull request by michnovka**

### CODE REVIEW

**Comment by michnovka** [comment]
@kukulich I dont think I can really improve the indentation efficiently. There is just not "correct" way of doing it, this should be enforced using other sniffs. What further complicates things is that we register the sniff on `T_ATTRIBUTE` so when I fix one attribute, I have now inconsistent actual fixed file with tokens from phpcs, as those point to the original parsed file.

I do not have more time to work on this, I am happy with the sniff as is. If you have some more suggestions, let me know.

**Comment by kukulich** [comment]
@michnovka I really don't understand why the sniff is done differently than suggested in https://github.com/slevomat/coding-standard/issues/1442#issuecomment-1276112153

**Comment by michnovka** [comment]
when implementing this I was facing some difficulties. Manily due to fixer and the fact this sniff targets `T_ATTRIBUTE`. By the time I fixed one attribute (e.g. add new line after it) it would modify the file for the 2nd one following and suddenly the tokens from phpcsfile were outdated (not corresponding to the fix-in-progress file). I couldnt make it work, so I asked myself the question - whats is the main purpose of this sniff? And that is to not allow having multiple attributes on one line. So thats what I implemented.

I am happy with renaming the sniff to `RequireAttributeOnDedicatedLine` but without targetting directly `AttributeHelper::attributableTokens` instead of T_ATTRIBUTEs I have hard time making it work as in the original comment.

**Comment by kukulich** [comment]
> I am happy with renaming the sniff to RequireAttributeOnDedicatedLine

I'm not. That's not what I suggested.

**Comment by michnovka** [comment]
Another issue is mixing of phpdoc and attributes. How to handle spacing there in regards to `linesCountBetweenAttributes`?

This sniff makes sure that attributes are always the only thing on a line. There is an exception for single in-line attributes

**Comment by michnovka** [comment]
@kukulich I have edited this PR based on https://github.com/slevomat/coding-standard/issues/1442#issuecomment-1279341581 Please review

File: SlevomatCodingStandard/Sniffs/Attributes/DisallowMultipleAttributesPerLineSniff.php [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+				self::CODE_ATTRIBUTE_NOT_ALONE_ON_LINE
+			);
+		}
+
+		if (
+			$tokens[$nextNonWhitespaceAfterAttributeCloserPointer]['code'] !== T_ATTRIBUTE &&
+			$tokens[$attributeCloserPointer]['line'] === $tokens[$nextNonWhitespaceAfterAttributeCloserPointer]['line']
+		) {
+			$fixNotAloneOnLine = $phpcsFile->addFixableError(
+				'Attribute is not alone on line',
```
Comment (kukulich) [inline SlevomatCodingStandard/Sniffs/Attributes/DisallowMultipleAttributesPerLineSniff.php:64]: `Multiple attributes per line are disallowed.`

**Review by kukulich — CHANGES_REQUESTED** [review]

File: tests/Sniffs/Attributes/data/disallowedMultipleAttributesPerLineErrors.fixed.php [hunk 2]
```diff
@@ -0,0 +1,101 @@
+<?php // lint >= 8.1
+
+#[Attribute1]
+#[Attribute2('var')]
+#[Attribute3(option: PDO::class, option2: true, option3: 'False')]
+#[Attribute4()]
+#[Attribute5]
+// comment
```
[the remark this answers is not in the document]
Comment (michnovka) [reply tests/Sniffs/Attributes/data/disallowedMultipleAttributesPerLineErrors.fixed.php:8]: @kukulich is there some helper function similar to `TokenHelper::findFirstNonWhitespaceOnLine()` that would look for non-whitespace AND non-comment?

[the remark this answers is not in the document]
Comment (kukulich) [reply tests/Sniffs/Attributes/data/disallowedMultipleAttributesPerLineErrors.fixed.php:8]: You can just check if there's another `T_ATTRIBUTE` on the same line before current attribute.

[the remark this answers is not in the document]
Comment (michnovka) [reply tests/Sniffs/Attributes/data/disallowedMultipleAttributesPerLineErrors.fixed.php:8]: no, I cannot. imagine:

```php
#[Attribute1]
#[Attribute2] /** whatever */ public int $x;
```

[the remark this answers is not in the document]
Comment (kukulich) [reply tests/Sniffs/Attributes/data/disallowedMultipleAttributesPerLineErrors.fixed.php:8]: I don't see any problem.

[the remark this answers is not in the document]
Comment (michnovka) [reply tests/Sniffs/Attributes/data/disallowedMultipleAttributesPerLineErrors.fixed.php:8]: it should be fixed into
```php
#[Attribute1]
#[Attribute2] /** whatever */
public int $x;
```

the point of this sniff is to have on one line ONLY one attribute AND whitespaces OR comments. But OK, we can make this sniff ONLY about 2 attributes on one line and make another sniff which disallows having attribute and attributed token on the same line. WDYT?

[the remark this answers is not in the document]
Comment (kukulich) [reply tests/Sniffs/Attributes/data/disallowedMultipleAttributesPerLineErrors.fixed.php:8]: > make another sniff which disallows having attribute and attributed token on the same line. WDYT?

That's the sniff that should solve number of lines between last attribute and owner (or comment).

[the remark this answers is not in the document]
Comment (michnovka) [reply tests/Sniffs/Attributes/data/disallowedMultipleAttributesPerLineErrors.fixed.php:8]: > That's the sniff that should solve number of lines between last attribute and owner (or comment).
How would you call that sniff?

[the remark this answers is not in the document]
Comment (kukulich) [reply tests/Sniffs/Attributes/data/disallowedMultipleAttributesPerLineErrors.fixed.php:8]: https://github.com/slevomat/coding-standard/commit/3b165784080535b827f6fa5c311d23d64fe38b7f

**Review by kukulich — APPROVED** [review]

**Comment by kukulich** [comment]
@michnovka Thanks! I've finished the sniff by myself.

**Pull request merged**